std/variant.d

```diff
@@ -472,6 +472,14 @@ private:
                 auto rhsPA = getPtr(&temp.store);
                 return compare(rhsPA, zis, selector);
             }
+            // Generate the function bellow only if the Variant's type is
+            // comparable with 'null'
+            static if (__traits(compiles, () => A.init == null))
```

Comment:
I'm not sure this is ok. I can imagine a variant with a struct that defines an alias this to a pointer. I don't know if we want to support that case, but if we do, this condition is not going to make it. A less restrictive condition could to simply see if the actual value can be tested for equality with null.

Also, note that this path is taken for comparison also. What happens if we have something along the lines of `a < null` (where a is a variant that stores a pointer). I would suggest that this should be an error (as it currently is).

Reply:
I'm trying to reproduce all behaviors that exist in the language. A Variant can hold any type and it "behaves" like the static D type that it currently holds. If `opEquals` is defined to interact directly with the underlying type then all valid operations that can be done with that type using the `==` operator should be reproducible in Variant.
```d
struct Foo
{
	int* ptr;
	alias ptr this;
}

void main()
{
	assert(Foo.init == null);
}
```

If this can be done with the core language, then IMO it should be possible with Variant.

> Also, note that this path is taken for comparison also.
I did not think of that. I think it's possible to restrict the logic to the `==` operator only if I'm not mistaken. I will fix this and add tests for it as well.

Also, comparing with `null` other than using `==` does not result in an error. It evaluates to `false` if `<` and to true or false if `<=, >=, ==, >`
```d
void main()
{
	int[] arr;
	assert(arr == null); // true
	assert(arr >= null); // true
	assert(arr <= null); // true
	assert(arr > null); // false
	assert(arr < null); // false
}
```

```d
void main()
{
	int[] arr = [3];
	assert(arr == null); // false
	assert(arr >= null); // true
	assert(arr <= null); // false
	assert(arr > null); // true
	assert(arr < null); // false
}
```

So I guess this should reproduce this logic.

Reply:
Makes sense. I'm on board.

Reply:
> I'm not sure this is ok. I can imagine a variant with a struct that defines an alias this to a pointer. I don't know if we want to support that case, but if we do, this condition is not going to make it. A less restrictive condition could to simply see if the actual value can be tested for equality with null.

I do not understand why this condition is not sufficient enough for that. Could you please clarify what you mean?

Reply:
Sorry, I got confused, don't mind that comment. I was also thinking of potential types that may accept `null` as a value but have a non-null init state. I don't know if any such types may exist, though.

Reply:
Well, weirdly enough the implemented logic already works as expected. I was tripping a bit but after checking the documentation for [opCmp](https://dlang.org/spec/operatoroverloading.html#compare) once again it made perfect sense. The comparison is rewritten to always compare with '0' and since my function always returns `0` or `1` this works.
* if comparing with `<` then it won't be `==` to `null` so the function returns `1` which correctly fails `opCmp`.
* if comparing with `<=` then it will correctly pass if the value is `null` fail otherwise for the same reason above.
* if comparing with `>, >=` it follows the normal convention of `opCmp` by returning `1` or `0`.

So I guess I don't need to change anything, but I should probably document the code for future comprehension. What do you say?

I will still add the unit tests for these cases.

Reply:
Yes, please. Fix the indenting, add the comment and we're golden.

Reply:
done